### PR TITLE
Add semantic for reading thread count from $OMP_NUM_THREADS

### PR DESCRIPTION
**Pull request by antoine-morvan**
Hello, 

I am using some scripts to automatically launch the CloudSC dwarf with various numbers of OpenMP thread. The framework however relies on `$OMP_NUM_THREAD` to set this value, and cannot act on the executable arguments.

This PR does 2 things:
1. Prevent the application from crashing when the first argument is 0 or negative, by adding the semantic below. Previously it was failing on trying to start regions with 0 threads, or allocating arrays of size -1 (leading to some funny stuff with some compilers and -Ofast ...);
2. Add semantic to the value of the first argument when zero or negative: it now reads the `$OMP_NUM_THREADS` variable from the environment if present or defaults to CPU count (`omp_get_max_threads()`);

The updated readme should reflect these changes.

This should not break existing behavior, as zero or negative arguments were leading to aplication crash.

Best regards.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf-ifs/dwarf-p-cloudsc?pullRequest=73) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf-ifs/dwarf-p-cloudsc?pullRequest=73) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf-ifs/dwarf-p-cloudsc?pullRequest=73) it.</sub>

File: src/cloudsc_c/dwarf_cloudsc.c [hunk 1]
```diff
@@ -33,6 +33,9 @@ int main( int argc, char *argv[] ) {
     omp_threads = atoi( argv[1] );
     ngptot      = atoi( argv[2] );
     nproma      = atoi( argv[3] );
+    if (omp_threads <= 0) {
+      omp_threads = omp_get_max_threads();
+    }
```
Comment (reuterbal) [inline src/cloudsc_c/dwarf_cloudsc.c:38]: Can we guard this with an `#ifdef _OPENMP` to retain the ability of building without OpenMP?

File: src/cloudsc_cuda/dwarf_cloudsc.cpp [hunk 2]
```diff
@@ -33,6 +33,9 @@ int main( int argc, char *argv[] ) {
     omp_threads = atoi( argv[1] );
     ngptot      = atoi( argv[2] );
     nproma      = atoi( argv[3] );
+    if (omp_threads <= 0) {
+      omp_threads = omp_get_max_threads();
+    }
```
Comment (reuterbal) [inline src/cloudsc_cuda/dwarf_cloudsc.cpp:38]: Also, guard with `#ifdef _OPENMP`?

File: src/cloudsc_fortran/dwarf_cloudsc.F90 [hunk 3]
```diff
@@ -19,6 +19,8 @@ PROGRAM DWARF_CLOUDSC
 USE YOMCST   , ONLY : YRCST 
 USE YOETHF   , ONLY : YRTHF
 
+USE OMP_LIB
+
```
Comment (reuterbal) [inline src/cloudsc_fortran/dwarf_cloudsc.F90:23]: Also, guard with `#ifdef _OPENMP`, and in l.53 below

File: src/cloudsc_gpu/dwarf_cloudsc_gpu.F90 [hunk 4]
```diff
@@ -42,6 +42,8 @@ PROGRAM DWARF_CLOUDSC
 USE CLOUDSC_DRIVER_GPU_SCC_FIELD_MOD, ONLY: CLOUDSC_DRIVER_GPU_SCC_FIELD
 #endif
 
+USE OMP_LIB
```
Comment (reuterbal) [inline src/cloudsc_gpu/dwarf_cloudsc_gpu.F90:45]: Guard with `#ifdef _OPENMP`, and in l.75 below

File: src/cloudsc_loki/dwarf_cloudsc.F90 [hunk 5]
```diff
@@ -18,6 +18,7 @@ PROGRAM DWARF_CLOUDSC
 USE CLOUDSC_DRIVER_LOKI_MOD, ONLY: CLOUDSC_DRIVER
 #endif
 USE EC_PMON_MOD, ONLY: EC_PMON
+USE OMP_LIB
```
Comment (reuterbal) [inline src/cloudsc_loki/dwarf_cloudsc.F90:21]: Also, guard with `#ifdef _OPENMP`, and in l.51 below

**Review by reuterbal — CHANGES_REQUESTED** [review]
Thanks, this looks like a good idea!

NB: I have retargeted the PR to develop

I haven't tested this myself, yet, but leaving a request for guarding the use of the OpenMP API, to make sure this can build without OpenMP available.

**Review by mlange05 — COMMENTED** [review]
Hi @antoine-morvan many thanks for the contribution. 

I also agree that this is a good idea, but that we might also need the `#ifdef` guards. However, I've run this and recreated the previous failure, and it all works nicely and as expected, so should be good to go once this is in place.

**Comment by antoine-morvan** [comment]
> Hi @antoine-morvan many thanks for the contribution.
> 
> I also agree that this is a good idea, but that we might also need the `#ifdef` guards. However, I've run this and recreated the previous failure, and it all works nicely and as expected, so should be good to go once this is in place.

Regarding the `#ifdef` in the C and CUDA implementations, the `cloudsc_drivers.h` files already includes `#include <omp.h>` without guard. The guard can be added nonetheless.

For the in Fortran files that would match what's in the `timer_mod`.  I'll add this.

File: src/cloudsc_c/dwarf_cloudsc.c (hunk 1 above)
Comment (antoine-morvan) [reply src/cloudsc_c/dwarf_cloudsc.c]: resolved in next commit

File: src/cloudsc_cuda/dwarf_cloudsc.cpp (hunk 2 above)
Comment (antoine-morvan) [reply src/cloudsc_cuda/dwarf_cloudsc.cpp]: resolved in next commit

File: src/cloudsc_fortran/dwarf_cloudsc.F90 (hunk 3 above)
Comment (antoine-morvan) [reply src/cloudsc_fortran/dwarf_cloudsc.F90]: resolved in next commit

File: src/cloudsc_gpu/dwarf_cloudsc_gpu.F90 (hunk 4 above)
Comment (antoine-morvan) [reply src/cloudsc_gpu/dwarf_cloudsc_gpu.F90:46]: resolved in next commit

File: src/cloudsc_loki/dwarf_cloudsc.F90 (hunk 5 above)
Comment (antoine-morvan) [reply src/cloudsc_loki/dwarf_cloudsc.F90:23]: resolved in next commit

**Comment by antoine-morvan** [comment]
Should be better with OpenMP-less support :)

**Review by mlange05 — APPROVED** [review]
Ok, fantastic. Many thanks @antoine-morvan  GTG from me.

**Review by reuterbal — APPROVED** [review]
Many thanks for adding the guards! We should sort this for CUDA/C as well, but that's for a different PR.

**Pull request merged**